File: Input/SFBInputSource.m
/*
 * Copyright (c) 2010 - 2020 Stephen F. Booth <me@sbooth.org>
 * See https://github.com/sbooth/SFBAudioEngine/blob/master/LICENSE.txt for license information
 */

#import "SFBInputSource.h"
#import "SFBInputSource+Internal.h"

#import "SFBDataInputSource.h"
#import "SFBFileContentsInputSource.h"
#import "SFBFileInputSource.h"
#import "SFBHTTPInputSource.h"
#import "SFBMemoryMappedFileInputSource.h"

os_log_t gSFBInputSourceLog = NULL;

static void SFBCreateInputSourceLog(void) __attribute__ ((constructor));
static void SFBCreateInputSourceLog()
{
	static dispatch_once_t onceToken;
	dispatch_once(&onceToken, ^{
		gSFBInputSourceLog = os_log_create("org.sbooth.AudioEngine", "InputSource");
	});
}

@implementation SFBInputSource

+ (instancetype)inputSourceForURL:(NSURL *)url flags:(SFBInputSourceFlags)flags error:(NSError **)error
{
	NSParameterAssert(url != nil);

	if(url.isFileURL) {
		if(flags & SFBInputSourceFlagsMemoryMapFiles)
			return [[SFBMemoryMappedFileInputSource alloc] initWithURL:url error:error];
		else if(flags & SFBInputSourceFlagsLoadFilesInMemory)
			return [[SFBFileContentsInputSource alloc] initWithContentsOfURL:url error:error];
		else
			return [[SFBFileInputSource alloc] initWithURL:url error:error];
	}
	else if([url.scheme.lowercaseString hasPrefix:@"http"])
		return [[SFBHTTPInputSource alloc] initWithURL:url error:error];

	return nil;
}

+ (instancetype)inputSourceWithData:(NSData *)data error:(NSError **)error
{
	NSParameterAssert(data != nil);
	return [[SFBDataInputSource alloc] initWithData:data];
}

+ (instancetype)inputSourceWithBytes:(const void *)bytes length:(NSInteger)length error:(NSError **)error
{
	NSParameterAssert(bytes != NULL);
	NSParameterAssert(length >= 0);
	return [[SFBDataInputSource alloc] initWithBytes:bytes length:length];
}

+ (instancetype)inputSourceWithBytesNoCopy:(void *)bytes length:(NSInteger)length freeWhenDone:(BOOL)freeWhenDone error:(NSError **)error
{
	NSParameterAssert(bytes != NULL);
	NSParameterAssert(length >= 0);
	return [[SFBDataInputSource alloc] initWithBytesNoCopy:bytes length:length freeWhenDone:freeWhenDone];
}

- (BOOL)openReturningError:(NSError **)error
{
	[self doesNotRecognizeSelector:_cmd];
	__builtin_unreachable();
}

- (BOOL)closeReturningError:(NSError **)error
{
	[self doesNotRecognizeSelector:_cmd];
	__builtin_unreachable();
}

- (BOOL)isOpen
{
	[self doesNotRecognizeSelector:_cmd];
	__builtin_unreachable();
}

- (BOOL)readBytes:(void *)buffer length:(NSInteger)length bytesRead:(NSInteger *)bytesRead error:(NSError **)error
{
	[self doesNotRecognizeSelector:_cmd];
	__builtin_unreachable();
}

- (BOOL)getOffset:(NSInteger *)offset error:(NSError **)error
{
	[self doesNotRecognizeSelector:_cmd];
	__builtin_unreachable();
}

- (BOOL)getLength:(NSInteger *)length error:(NSError **)error
{
	[self doesNotRecognizeSelector:_cmd];
	__builtin_unreachable();
}

- (BOOL)supportsSeeking
{
	[self doesNotRecognizeSelector:_cmd];
	__builtin_unreachable();
}

- (BOOL)seekToOffset:(NSInteger)offset error:(NSError **)error
{
	[self doesNotRecognizeSelector:_cmd];
	__builtin_unreachable();
}

@end

@implementation SFBInputSource (SFBSignedIntegerReading)
- (BOOL)readInt8:(int8_t *)i8 error:(NSError **)error			{ return [self readUInt8:(uint8_t *)i8 error:error]; }
- (BOOL)readInt16:(int16_t *)i16 error:(NSError **)error		{ return [self readUInt16:(uint16_t *)i16 error:error]; }
- (BOOL)readInt32:(int32_t *)i32 error:(NSError **)error		{ return [self readUInt32:(uint32_t *)i32 error:error]; }
- (BOOL)readInt64:(int64_t *)i64 error:(NSError **)error		{ return [self readUInt64:(uint64_t *)i64 error:error]; }
@end

@implementation SFBInputSource (SFBUnsignedIntegerReading)
- (BOOL)readUInt8:(uint8_t *)ui8 error:(NSError **)error
{
	NSInteger bytesRead;
	return [self readBytes:ui8 length:sizeof(uint8_t) bytesRead:&bytesRead error:error] && bytesRead == sizeof(uint8_t);
}

- (BOOL)readUInt16:(uint16_t *)ui16 error:(NSError **)error
{
	NSInteger bytesRead;
	return [self readBytes:ui16 length:sizeof(uint16_t) bytesRead:&bytesRead error:error] && bytesRead == sizeof(uint16_t);
}

- (BOOL)readUInt32:(uint32_t *)ui32 error:(NSError **)error
{
	NSInteger bytesRead;
	return [self readBytes:ui32 length:sizeof(uint32_t) bytesRead:&bytesRead error:error] && bytesRead == sizeof(uint32_t);
}

- (BOOL)readUInt64:(uint64_t *)ui64 error:(NSError **)error
{
	NSInteger bytesRead;
	return [self readBytes:ui64 length:sizeof(uint64_t) bytesRead:&bytesRead error:error] && bytesRead == sizeof(uint64_t);
}

@end

@implementation SFBInputSource (SFBBigEndianReading)

- (BOOL)readUInt16BigEndian:(uint16_t *)ui16 error:(NSError **)error
{
	NSParameterAssert(ui16 != nil);
	if(![self readUInt16:ui16 error:error])
		return NO;
	*ui16 = OSSwapHostToBigInt16(*ui16);
	return YES;
}

- (BOOL)readUInt32BigEndian:(uint32_t *)ui32 error:(NSError **)error
{
	NSParameterAssert(ui32 != nil);
	if(![self readUInt32:ui32 error:error])
		return NO;
	*ui32 = OSSwapHostToBigInt32(*ui32);
	return YES;
}

- (BOOL)readUInt64BigEndian:(uint64_t *)ui64 error:(NSError **)error
{
	NSParameterAssert(ui64 != nil);
	if(![self readUInt64:ui64 error:error])
		return NO;
	*ui64 = OSSwapHostToBigInt64(*ui64);
	return YES;
}

@end

@implementation SFBInputSource (SFBLittleEndianReading)

- (BOOL)readUInt16LittleEndian:(uint16_t *)ui16 error:(NSError **)error
{
	NSParameterAssert(ui16 != nil);
	if(![self readUInt16:ui16 error:error])
		return NO;
	*ui16 = OSSwapHostToLittleInt16(*ui16);
	return YES;
}

- (BOOL)readUInt32LittleEndian:(uint32_t *)ui32 error:(NSError **)error
{
	NSParameterAssert(ui32 != nil);
	if(![self readUInt32:ui32 error:error])
		return NO;
	*ui32 = OSSwapHostToLittleInt32(*ui32);
	return YES;
}

- (BOOL)readUInt64LittleEndian:(uint64_t *)ui64 error:(NSError **)error
{
	NSParameterAssert(ui64 != nil);
	if(![self readUInt64:ui64 error:error])
		return NO;
	*ui64 = OSSwapHostToLittleInt64(*ui64);
	return YES;
}

@end

File: Input/SFBInputSource.swift
/*
 * Copyright (c) 2020 Stephen F. Booth <me@sbooth.org>
 * See https://github.com/sbooth/SFBAudioEngine/blob/master/LICENSE.txt for license information
 */

import Foundation

extension InputSource {
	/// Reads bytes from the input
	/// - parameter buffer: A buffer to receive data
	/// - parameter length: The maximum number of bytes to read
	/// - returns: The number of bytes actually read
	/// - throws: An `NSError` object if an error occurs
	public func read(_ buffer: UnsafeMutableRawPointer, length: Int) throws -> Int {
		var bytesRead = 0
		try __readBytes(buffer, length: length, bytesRead: &bytesRead)
		return bytesRead
	}

	/// Returns the current offset in the input, in bytes
	/// - throws: An `NSError` object if an error occurs
	public func offset() throws -> Int {
		var offset = 0
		try __getOffset(&offset)
		return offset
	}

	/// Returns the length of the input, in bytes
	/// - throws: An `NSError` object if an error occurs
	public func length() throws -> Int {
		var length = 0
		try __getLength(&length)
		return length
	}

	/// Reads and returns a binary integer
	/// - returns: The integer value read
	/// - throws: An `NSError` object if an error occurs
	public func read<T: BinaryInteger>() throws -> T {
		var i: T = 0
		var bytesRead = 0

		let size = MemoryLayout<T>.size

		try withUnsafePointer(to: &i) {
			bytesRead = try read(UnsafeMutablePointer(mutating: $0), length: size)
		}

		if bytesRead != size {
			throw NSError(domain: NSPOSIXErrorDomain, code: Int(EIO), userInfo: nil)
		}

		return i
	}

	/// Reads a 16-bit integer from the input in big-endian format and returns the value
	/// - throws: An `NSError` object if an error occurs
	public func readBigEndian() throws -> UInt16 {
		let ui16: UInt16 = try read()
		return CFSwapInt16HostToBig(ui16)
	}

	/// Reads a 32-bit integer from the input in big-endian format and returns the value
	/// - throws: An `NSError` object if an error occurs
	public func readBigEndian() throws -> UInt32 {
		let ui32: UInt32 = try read()
		return CFSwapInt32HostToBig(ui32)
	}

	/// Reads a 64-bit integer from the input in big-endian format and returns the value
	/// - throws: An `NSError` object if an error occurs
	public func readBigEndian() throws -> UInt64 {
		let ui64: UInt64 = try read()
		return CFSwapInt64HostToBig(ui64)
	}

	/// Reads a 16-bit integer from the input in little-endian format and returns the value
	/// - throws: An `NSError` object if an error occurs
	public func readLittleEndian() throws -> UInt16 {
		let ui16: UInt16 = try read()
		return CFSwapInt16HostToLittle(ui16)
	}

	/// Reads a 32-bit integer from the input in little-endian format and returns the value
	/// - throws: An `NSError` object if an error occurs
	public func readLittleEndian() throws -> UInt32 {
		let ui32: UInt32 = try read()
		return CFSwapInt32HostToLittle(ui32)
	}

	/// Reads a 64-bit integer from the input in little-endian format and returns the value
	/// - throws: An `NSError` object if an error occurs
	public func readLittleEndian() throws -> UInt64 {
		let ui64: UInt64 = try read()
		return CFSwapInt64HostToLittle(ui64)
	}

}
